### ISSUE TITLE
Single letter strings get parsed as italicized in math mode

### ISSUE THREAD

**Issue by SekoiaTree** [typst/typst#274]
```
#let q1="q"
#let q2="qq"
$ q q1 q2 "qq" $
```

![image](https://user-images.githubusercontent.com/51149447/227609317-90dc4f6b-15b2-43f7-8119-b4f7f54a55d0.png)

It seems strange to me that `"q"` is rendered the same as `q`. Currently, this can technically be worked around with `"q "`, but that adds a space (not particularly different, but still slightly)

**Comment by jakobjpeters** [comment]
I agree that it is unexpected and inconsistent. Since there is already a way to create the italic character (no quotes), it would be better to make characters inside quotes behave consistently.

**Comment by GaugeAndGravity** [comment]
Find the same issue when I was trying `#let dd="d"` for differential and felt surprised on this behavior...

**Comment by dccsillag** [comment]
> Find the same issue when I was trying `#let dd="d"` for differential and felt surprised on this behavior...

BTW, for differential you can just use `dif`.

**Comment by mattfbacon** [comment]
FYI https://github.com/typst/typst/blob/main/library/src/math/ctx.rs#L132-L146

**Comment by thinety** [comment]
I would like to add that I was also surprised by this behavior when trying to write units that use special symbols / Greek letters: `$R = 50 med "k" Omega$` displays the `k` in italic. The alternatives are `$R = 50 med upright(k) Omega$` or
```
#let kohm = "k" + sym.Omega
$R = 50 med kohm$
```

**Comment by lucasvreis** [comment]
Looking at this and #366, I think the current loose and automatic separation between "math variable" and "prose within math" is problematic, especially if you consider that not everybody speaks English. In a plethora of languages you have single-letter connectives like "e" and "y" ("and" in Portuguese and Spanish), and they should appear inside maths upright. But I'm very surprised that _none_ of the following work:
```
$ 1+1 = 2 "e" 1+2=3 $ // Italic
$ 1+1 = 2 text("e") 1+2=3 $ // Italic
$ 1+1 = 2 upright("e") 1+2=3 $ // Wrong spacing!!
$ 1+1 = 2 upright(" e ") 1+2=3 $ // Wrong spacing again!!!
$ 1+1 = 2upright(" e ")1+2=3 $ // Wrong spacing!!!
```
The closest seems to be this ugly form:
```
$ 1+1 = 2upright(" e") 1+2=3 $ // finally!
```
If "multi letter variables" are a thing, I think they should have a different syntax from "text within maths" because semantically they are very different things, and having the same syntax for both is prone to produce such typographic inconsistencies.

**Comment by binarynoise** [comment]
If we said, ok, text in quotes is upright, text without quotes is italic (or vice versa) and text with a `#` prepended is a variable.

This would also improve the two-letter unit parsed as variable situation where it at least still compiles but just looks wrong instead of refusing to compile.

**Comment by ModProg** [comment]
> If we said, ok, text in quotes is upright, text without quotes is italic (or vice versa) and text with a `#` prepended is a variable, this would clarify things and also improve the two-letter unit parsed as variable situation where it at least still compiles but just looks wrong instead of refusing to compile.

I agree with the first half, but having the non `#` variables in math mode are quite nice if you have a bunch of `arg max sum ...`

**Comment by Enter-tainer** [comment]
A more recent pointer https://github.com/typst/typst/blob/ffc9570c643540282d35dfe3075567c1200e4171/crates/typst/src/math/ctx.rs#L193

**Comment by Leedehai** [comment]
It looks like a bug, but curiously the current behavior does not break what the doc (https://typst.app/docs/reference/math/) says:

```
In math, single letters are always displayed as is. Multiple letters,
however, are interpreted as variables and functions. To display multiple letters
verbatim, you can place them into quotes and to access
single letter variables, you can use the hash syntax.
```

If we want to display quoted single letter as normal text too, as suggested in this issue (and I agree!), Typst should promise this instead:
```
In math, single letters are always displayed as is. Multiple letters,
however, are interpreted as variables and functions. To display letters
verbatim like normal text, you can place them into quotes; to access
single-letter variables, you can use the hash syntax.
```

So.. do we want this? Bringing to @laurmaedje 's attention.

**Comment by Leedehai** [comment]
> One idea is that during the math eval phase... this direction doesn't look right.

Hmm, I seem to have made it work by carving out a special case during the math eval phase's `eval_display()`

```rs
impl ExprExt for ast::Expr<'_> {
    fn eval_display(&self, vm: &mut Vm) -> SourceResult<Content> {
        let value = self.eval(vm)?;
        let content = match value {
            Value::Str(_) => {
                value.display().styled(EquationElem::set_italic(Smart::Custom(false)))
            }
            _ => value.display(),
        };
        Ok(content.spanned(self.span()))
    }
}
```
Result:
```typ
$ 1 "ft" = 0.3048 "m" \
  A = pi r^2 $
```
<img width="215" src="https://github.com/typst/typst/assets/18319900/43289eef-f93b-4464-972f-e206639f04c9">

**Comment by laurmaedje** [comment]
@Leedehai Forcing upright at that point would unfortunately break the `italic` function:
```typ
$ 1 "ft" = 0.3048 "m" \
  A = pi r^2 italic(a + "beta") $
```
(Just `italic("beta")` continues to work because it doesn't use that code path, which would add extra inconsistency.)

I consider this issue a bug or at least something that's not desirable, but I'm not sure whether we can properly fix it without something like #1125.

**Comment by Leedehai** [comment]
Agreed.

Further, special-casing `Value::Str` could over-apply the italic-to-upright fix, because it turned out the function parameter names are collected as `Value::Str` into the argument sink, too, and thus when they are embedded into `$..$` they will be in upright font, which is a mistake.

Barring an overhaul like #1125, it looks an appropriate limited-in-scope fix should start from the parsing phase. The knowledge of different Value types in the evaluation phase is not sufficient to implement this fix.

**Comment by mattfbacon** [comment]
Yes, I think there could even be added a new element type `math.lit` for these literals which could even facilitate special styling or overriding the default upright style.

**Comment by gasche** [comment]
I also hit this issue: I would like `"Foo"` and `"F"` to generate the same `F` character in math context, and right now it looks like we have to write `upright(F)` instead of `"F"` to get this result.

My understanding from reading the issue is that this is recognized as a bug, but one that is not so easy to fix. The PR #4638 from @wrzian was submitted just this summer and claims to fix the issue, but it hasn't been reviewed yet. (It is marked in "draft" mode and @wrzian wants to write more documentation.) From a distance, it looks like beginners who are not familiar with the Typst codebase cannot really do much to help currently.

**Comment by jtchen2k** [comment]
I also found that this issue made the output of `bold(x)` inconsistent with LaTeX's `\mathbf{x}`. Typst gives a bold, italic x, which is unexpected since we want it to be upright to denote a vector 𝐱. I am currently using a workaround like `#let bf(body) = math.bold(math.upright(body))`.
